### PR TITLE
fix: fixed title of contact doctype and added new field for title

### PR DESCRIPTION
**Pull request by meshramsaroj**
**Task Link:** https://app.asana.com/0/1199639423748061/1199955181219855/f

**Description:** Added title field into contact doctype. Fix contact's title behavior, now the title will update as per changes. also made name value equal to CONTACT00001

**Screenshot:** 

- Before
image

- After
updated_contact

### CODE REVIEW

File: frappe/contacts/doctype/contact/contact.json [hunk 1]
```diff
@@ -2,6 +2,7 @@
  "allow_events_in_timeline": 1,
  "allow_import": 1,
  "allow_rename": 1,
+ "autoname": "CONTACT.#####",
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.json]: - autoname should look like `Contact-0001`

File: frappe/contacts/doctype/contact/contact.py [hunk 2]
```diff
@@ -122,6 +122,17 @@ def set_primary(self, fieldname):
 				setattr(self, fieldname, d.phone)
 				break
 
+	def before_save(self):
+		self.title = f'{self.first_name} {self.last_name or ""}'
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]:
```suggestion
		self.title = "{0} {1}".format(self.first_name, self.last_name)
```

File: frappe/contacts/doctype/contact/contact.py [hunk 3]
```diff
@@ -122,6 +122,17 @@ def set_primary(self, fieldname):
...
+
+		if frappe.db.exists("Contact", self.title):
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]: - self.title is not the actual name!
```suggestion
		if frappe.db.exists("Contact", {"title": self.title}):
```

File: frappe/contacts/doctype/contact/contact.py [hunk 4]
```diff
@@ -17,8 +17,8 @@
 class Contact(Document):
 	def autoname(self):
 		# concat first and last name
-		self.name = " ".join(filter(None,
-			[cstr(self.get(f)).strip() for f in ["first_name", "last_name"]]))
+		# self.name = " ".join(filter(None,
+		# 	[cstr(self.get(f)).strip() for f in ["first_name", "last_name"]]))
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]: Remove the commented code

**Review by hrwX — DISMISSED** [review]

File: frappe/contacts/doctype/contact/contact.py (hunk 4 above)
Comment (meshramsaroj) [reply frappe/contacts/doctype/contact/contact.py]: removed commented code

File: frappe/contacts/doctype/contact/contact.py (hunk 3 above)
Comment (meshramsaroj) [reply frappe/contacts/doctype/contact/contact.py]: fixed this change

File: frappe/contacts/doctype/contact/contact.json (hunk 1 above)
Comment (meshramsaroj) [reply frappe/contacts/doctype/contact/contact.json]: made name format like Contact-0001
image

File: frappe/contacts/doctype/contact/contact.py (hunk 2 above)
Comment (meshramsaroj) [reply frappe/contacts/doctype/contact/contact.py]: fixed this change

File: frappe/contacts/doctype/contact/contact.py [hunk 5]
```diff
@@ -11,14 +11,14 @@
...
+from frappe.model.naming   import make_autoname
 
 import functools
 
 class Contact(Document):
 	def autoname(self):
 		# concat first and last name
-		self.name = " ".join(filter(None,
-			[cstr(self.get(f)).strip() for f in ["first_name", "last_name"]]))
+		self.name= make_autoname("Contact-.####")
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]:
```suggestion
		self.name = make_autoname("Contact-.####")
```

File: frappe/contacts/doctype/contact/contact.py [hunk 6]
```diff
@@ -122,6 +122,17 @@ def set_primary(self, fieldname):
 				setattr(self, fieldname, d.phone)
 				break
 
+	def before_save(self):
+		self.title = "{0} {1}".format(self.first_name, self.last_name)
+
+		if frappe.db.exists("Contact", {"title":self.title}):
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]:
```suggestion
		if frappe.db.exists("Contact", {"title": self.title}):
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: frappe/contacts/doctype/contact/contact.py [hunk 7]
```diff
@@ -11,14 +11,14 @@
 from past.builtins import cmp
 from frappe.model.naming import append_number_if_name_exists
 from frappe.contacts.address_and_contact import set_link_title
+from frappe.model.naming   import make_autoname
```
Comment (sahil28297) [inline frappe/contacts/doctype/contact/contact.py]:
```suggestion
from frappe.model.naming import make_autoname
```

**Review by sahil28297 — CHANGES_REQUESTED** [review]

**Review by sahil28297 — DISMISSED** [review]

File: frappe/contacts/doctype/contact/contact.json [hunk 8]
```diff
@@ -240,12 +241,18 @@
...
+   "fieldtype": "Data",
+   "hidden": 1,
+   "label": "title"
   }
  ],
  "icon": "fa fa-user",
  "idx": 1,
  "image_field": "image",
- "modified": "2019-10-10 22:04:41.070479",
+ "modified": "2021-03-18 15:44:20.711966",
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.json]: Enable `show_title_field_in_link` for Contact

**Review by hrwX — CHANGES_REQUESTED** [review]

File: frappe/contacts/doctype/contact/contact.json (hunk 8 above)
Comment (meshramsaroj) [reply frappe/contacts/doctype/contact/contact.json]: Enabled show_title_field_in_link field

**Review by hrwX — DISMISSED** [review]

**Comment by hrwX** [comment]
@meshramsaroj write a patch to update the new title fields

File: frappe/patches/v12_0/set_contact_title.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    print("===============this is patch=======")
+    frappe.reload_doc("Contacts", "doctype", "Contact")
+    contacts = frappe.get_all("Contact",["name","title"])
+    for contact in contacts:
+        frappe.db.sql("""
+                        UPDATE
+                            `tab%s`
+                        SET
+				            contact.title = contact.name
+            """% (contact))
```
Comment (hrwX) [inline frappe/patches/v12_0/set_contact_title.py]:
```suggestion
	frappe.reload_doc("contacts", "doctype", "contact")
	frappe.db.sql("""
		UPDATE `tabContact`
		SET title=name
	""")
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: frappe/contacts/doctype/contact/contact.py [hunk 10]
```diff
@@ -11,14 +11,14 @@
 from past.builtins import cmp
 from frappe.model.naming import append_number_if_name_exists
 from frappe.contacts.address_and_contact import set_link_title
+from frappe.model.naming import make_autoname
 
 import functools
 
 class Contact(Document):
 	def autoname(self):
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]: Remove this autoname and put this in doctype, refer [this](https://frappeframework.com/docs/user/en/basics/doctypes/naming)

File: frappe/contacts/doctype/contact/contact.py [hunk 11]
```diff
@@ -122,6 +122,17 @@ def set_primary(self, fieldname):
 				setattr(self, fieldname, d.phone)
 				break
 
+	def before_save(self):
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]: - Put this in validate

**Review by hrwX — CHANGES_REQUESTED** [review]

File: frappe/contacts/doctype/contact/contact.py [hunk 12]
```diff
@@ -16,10 +16,6 @@
 
 class Contact(Document):
 	def autoname(self):
-		# concat first and last name
-		self.name = " ".join(filter(None,
-			[cstr(self.get(f)).strip() for f in ["first_name", "last_name"]]))
-
 		if frappe.db.exists("Contact", self.name):
 			self.name = append_number_if_name_exists('Contact', self.name)
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]: Remove this autoname function completely

File: frappe/contacts/doctype/contact/contact.py [hunk 13]
```diff
@@ -122,6 +119,17 @@ def set_primary(self, fieldname):
 				setattr(self, fieldname, d.phone)
 				break
 
+	def before_save(self):
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]:
```suggestion
	def set_title(self):
```

File: frappe/contacts/doctype/contact/contact.py [hunk 14]
```diff
@@ -32,6 +28,7 @@ def validate(self):
 		self.set_primary_email()
 		self.set_primary("phone")
 		self.set_primary("mobile_no")
+		self.before_save()
```
Comment (hrwX) [inline frappe/contacts/doctype/contact/contact.py]:
```suggestion
		self.set_title()
```

**Review by hrwX — CHANGES_REQUESTED** [review]

**Review by hrwX — APPROVED** [review]

**Review by neilLasrado — APPROVED** [review]

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**